depends: Add RISC-V support

laanwj opened the pull request:
This adds support for riscv32 and riscv64 builds to the depends system. 

The change consists of documentation and build system changes. The most significant change is an update of `config.sub` and `config.guess` inside zeromq patch, as the current version does not recognize the `riscv*` host tuples (there's no new version of ZeroMQ yet with newer ones).

Good thing: RISC-V 64-bit toolchain packages can be installed out of the box on Ubuntu 18.04+.

I would also like to add RISC-V 64-bit executables to gitian, but this will not be possible until #12511 .

laanwj committed:
depends: Add RISC-V support

fanquake commented:
@laanwj Have you been building with `NO_QT=1`?

I've just tested a `make HOST=riscv64-linux-gnu ` depends build on Ubuntu 18.04, after installing the RISC-V toolchain and am seeing:
```
Configuring qrencode...
checking build system type... x86_64-unknown-linux-gnu
checking host system type... Invalid configuration `riscv64-linux-gnu': machine `riscv64' not recognized
configure: error: /bin/bash use/config.sub riscv64-linux-gnu failed
```
Building with `NO_QT=1` completes successfully.

I wonder if there's a less verbose patch we could et away with to fix the zeromq build?

laanwj commented:
Good point: yes, I have built without Qt, same as we do for ARM

tmagik commented:
I can test this on a HiFiveU board next week. The problem with running a full bitcoin main net instance on a physical riscv board is going to be I/O.

In the meantime, try running in Qemu as it will probably have better I/O

tmagik commented:
> Thanks for testing. Good point: yes, I have built without Qt, same as we do for ARM. Should probably mention that in the documentation, though I don't think there's any RISC-V boards at this time with a capable enough GPU and HDMI/VGA out to run even a desktop GUI (though maybe it's possible to connect one through the PCI-e slot? I don't know).

https://www.crowdsupply.com/microsemi/hifive-unleashed-expansion-board

It runs X with a radeon graphics card and has more cores/memory/performance than the first machine I compiled bitcoin on. Bitcoin ATM vendors looking at high volume, low-cost embedded systems should probably be spending money on some evaluation systems.

laanwj commented:
This is interesting:
```
user@bionic:~/bitcoin/src$ QEMU_LD_PREFIX=/usr/riscv64-linux-gnu qemu-riscv64 ./test/test_bitcoin
Running 291 test cases...
test/crypto_tests.cpp(561): error: in "crypto_tests/sha256d64": check memcmp(out1, out2, 32 * i) == 0 has failed
test/crypto_tests.cpp(561): error: in "crypto_tests/sha256d64": check memcmp(out1, out2, 32 * i) == 0 has failed
test/crypto_tests.cpp(561): error: in "crypto_tests/sha256d64": check memcmp(out1, out2, 32 * i) == 0 has failed
test/crypto_tests.cpp(561): error: in "crypto_tests/sha256d64": check memcmp(out1, out2, 32 * i) == 0 has failed
test/crypto_tests.cpp(561): error: in "crypto_tests/sha256d64": check memcmp(out1, out2, 32 * i) == 0 has failed
test/crypto_tests.cpp(561): error: in "crypto_tests/sha256d64": check memcmp(out1, out2, 32 * i) == 0 has failed
test/crypto_tests.cpp(561): error: in "crypto_tests/sha256d64": check memcmp(out1, out2, 32 * i) == 0 has failed
test/crypto_tests.cpp(561): error: in "crypto_tests/sha256d64": check memcmp(out1, out2, 32 * i) == 0 has failed
test/crypto_tests.cpp(561): error: in "crypto_tests/sha256d64": check memcmp(out1, out2, 32 * i) == 0 has failed
test/crypto_tests.cpp(561): error: in "crypto_tests/sha256d64": check memcmp(out1, out2, 32 * i) == 0 has failed
test/crypto_tests.cpp(561): error: in "crypto_tests/sha256d64": check memcmp(out1, out2, 32 * i) == 0 has failed
test/crypto_tests.cpp(561): error: in "crypto_tests/sha256d64": check memcmp(out1, out2, 32 * i) == 0 has failed
test/crypto_tests.cpp(561): error: in "crypto_tests/sha256d64": check memcmp(out1, out2, 32 * i) == 0 has failed
test/crypto_tests.cpp(561): error: in "crypto_tests/sha256d64": check memcmp(out1, out2, 32 * i) == 0 has failed
test/crypto_tests.cpp(561): error: in "crypto_tests/sha256d64": check memcmp(out1, out2, 32 * i) == 0 has failed
test/crypto_tests.cpp(561): error: in "crypto_tests/sha256d64": check memcmp(out1, out2, 32 * i) == 0 has failed
test/crypto_tests.cpp(561): error: in "crypto_tests/sha256d64": check memcmp(out1, out2, 32 * i) == 0 has failed
test/crypto_tests.cpp(561): error: in "crypto_tests/sha256d64": check memcmp(out1, out2, 32 * i) == 0 has failed
test/crypto_tests.cpp(561): error: in "crypto_tests/sha256d64": check memcmp(out1, out2, 32 * i) == 0 has failed
test/crypto_tests.cpp(561): error: in "crypto_tests/sha256d64": check memcmp(out1, out2, 32 * i) == 0 has failed
test/crypto_tests.cpp(561): error: in "crypto_tests/sha256d64": check memcmp(out1, out2, 32 * i) == 0 has failed
test/crypto_tests.cpp(561): error: in "crypto_tests/sha256d64": check memcmp(out1, out2, 32 * i) == 0 has failed
test/crypto_tests.cpp(561): error: in "crypto_tests/sha256d64": check memcmp(out1, out2, 32 * i) == 0 has failed
test/crypto_tests.cpp(561): error: in "crypto_tests/sha256d64": check memcmp(out1, out2, 32 * i) == 0 has failed
test/crypto_tests.cpp(561): error: in "crypto_tests/sha256d64": check memcmp(out1, out2, 32 * i) == 0 has failed
test/crypto_tests.cpp(561): error: in "crypto_tests/sha256d64": check memcmp(out1, out2, 32 * i) == 0 has failed
test/crypto_tests.cpp(561): error: in "crypto_tests/sha256d64": check memcmp(out1, out2, 32 * i) == 0 has failed
test/crypto_tests.cpp(561): error: in "crypto_tests/sha256d64": check memcmp(out1, out2, 32 * i) == 0 has failed
test/crypto_tests.cpp(561): error: in "crypto_tests/sha256d64": check memcmp(out1, out2, 32 * i) == 0 has failed
test/crypto_tests.cpp(561): error: in "crypto_tests/sha256d64": check memcmp(out1, out2, 32 * i) == 0 has failed
test/crypto_tests.cpp(561): error: in "crypto_tests/sha256d64": check memcmp(out1, out2, 32 * i) == 0 has failed
test/crypto_tests.cpp(561): error: in "crypto_tests/sha256d64": check memcmp(out1, out2, 32 * i) == 0 has failed

*** 32 failures are detected in the test module "Bitcoin Test Suite"
```

sipa commented:
@laanwj That's very strange. It should be using the pure C++ SHA256 implementation. What SHA256 implementation does debug.log report (if you get that far)?

laanwj commented:
> @laanwj That's very strange. It should be using the pure C++ SHA256 implementation. What SHA256 implementation does debug.log report (if you get that far)?

It chooses the correct one:
```
2018-06-27T19:16:24Z Using the 'standard' SHA256 implementation
```
I'll dive into happens here.

laanwj commented:
It looks like the beginning of input `in[]` to `SHA256D64` becomes corrupted by
```c++
        for (int j = 0; j < i; ++j) {
            CHash256().Write(in + 64 * j, 64).Finalize(out1 + 32 * j);
        }
```
The function itself works correctly.
I'm suspecting a miscompilation issue at the moment.

laanwj commented:
My sync in qemu is at block 270000, and has encountered no issues. I haven't figured out yet why the `crypto_tests/sha256d64` specifically has this issue. Strangely enough, moving the `in` array to the outer scope makes the test pass.
I intend to try compilng with a more recent version of gcc to see if the test problem is resolved.

tmagik commented:
I attempted to make a test case, and simplified the following down:
```
static inline uint64_t InsecureRandBits(int bits) { return insecure_seed++; }
```

however this does not fail, so I'm now highly suspicious of inlining

tmagik commented:
@laanwj Debian's gcc-8 appears to fix the problem
```
riscv64-linux-gnu-gcc-8 --version
riscv64-linux-gnu-gcc-8 (Debian 8.1.0-8) 8.1.0
```

laanwj commented:
@tmagic Thanks for reproducing. Happy to hear it's solved with newer gcc. It might be better to remove the note about using ubuntu's cross-toolchain packages, then, until they have gcc 8 like debian.

theuni commented on depends/packages/zeromq.mk:
```diff
@@ -3,7 +3,7 @@ $(package)_version=4.2.3
 $(package)_file_name=$(package)-$($(package)_version).tar.gz
 $(package)_sha256_hash=8f1e2b2aade4dbfde98d82366d61baef2f62e812530160d2e6d0a5bb24e40bc0
-$(package)_patches=0001-fix-build-with-older-mingw64.patch 0002-disable-pthread_set_name_np.patch
+$(package)_patches=0001-fix-build-with-older-mingw64.patch 0002-disable-pthread_set_name_np.patch 0003-new_configure.patch
```
Could we [copy our files](https://github.com/bitcoin/bitcoin/blob/master/depends/packages/bdb.mk#L18) instead, since we have to keep ours up to date anyway?

fanquake replied:
I've added some commits to a branch that attempt to do that: https://github.com/fanquake/bitcoin/tree/riscv-depends-no-zmq-patching. It also updates the depends config.guess/sub to latest upstream.

@laanwj You can cherry pick from there if you'd like. Note that branch doesn't include re-running `./autogen.sh` like below. I'll comment about that shortly.

theuni commented on depends/packages/zeromq.mk:
```diff
@@ -13,7 +13,9 @@ endef
-   patch -p1 < $($(package)_patch_dir)/0002-disable-pthread_set_name_np.patch
+   patch -p1 < $($(package)_patch_dir)/0002-disable-pthread_set_name_np.patch && \
+   patch -p1 < $($(package)_patch_dir)/0003-new_configure.patch && \
+   ./autogen.sh
```
Is it definitely necessary to re-run autogen.sh? Afaik configure just calls out to config.guess/config.sub. Unless there's some weird globbing involved, I should think that replacing those files would be enough.

theuni reviewed (approved):
utACK, suggested changes would be a bonus if they work.

fanquake committed:
depends: latest config.guess

fanquake committed:
depends: latest config.sub

fanquake committed:
depends: update zmq config.guess/config.sub for riscv support

fanquake commented:
I've done some more testing of this using a [branch](https://github.com/fanquake/bitcoin/tree/riscv-depends-no-zmq-patching) that doesn't patch zmq's config.guess/sub, and instead just copies over our local versions, which I've updated from upstream.

re running [`./autogen.sh` after copying](https://github.com/bitcoin/bitcoin/pull/13543#discussion_r200713627). 
With or without `'./autogen.sh` I see this warning from the zmq build:
```
Staging zeromq...
5.	make[1]: warning: jobserver unavailable: using -j1.  Add '+' to parent make rule.	5.	make[1]: warning: jobserver unavailable: using -j1.  Add '+' to parent make rule.
```
The other difference (at least in output) is:
With `./autogen.sh`:
```
libtool: warning: remember to run 'libtool --finish /home/ubuntu/bitcoin/depends/riscv64-linux-gnu/lib'
```

Without `./autogen.sh`:
```
libtool: install: warning: remember to run `libtool --finish /home/ubuntu/bitcoin/depends/riscv64-linux-gnu/lib'
```

Also noticed that when I pass `--enable-werror` to configure I see:
```
./configure --prefix=`pwd`/depends/riscv64-linux-gnu --enable-werror
<snip>

debug enabled = no
  gprof enabled = no
  werror        = yes

  target os     = linux
  build os      = 

  CC            = riscv64-linux-gnu-gcc
  CFLAGS        = -pipe -O2 
  CPPFLAGS      =   -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=2 -I/home/ubuntu/bitcoin/depends/riscv64-linux-gnu/share/../include/  -DHAVE_BUILD_INFO -D__STDC_FORMAT_MACROS
  CXX           = riscv64-linux-gnu-g++ -std=c++11
  CXXFLAGS      =   -Wstack-protector -fstack-protector-all    -Werror=vla  -pipe -O2 
  LDFLAGS       = -pthread  -Wl,-z,relro -Wl,-z,now -pie  -L/home/ubuntu/bitcoin/depends/riscv64-linux-gnu/share/../lib 
  ARFLAGS       = cr
```

```
  CXXLD    bench/bench_bitcoin
  CXXLD    test/test_bitcoin
/home/ubuntu/bitcoin/depends/riscv64-linux-gnu/share/../lib/libboost_unit_test_framework-mt.a(execution_monitor.o): In function `.L0 ':
execution_monitor.cpp:(.text+0x9a4): warning: fedisableexcept is not implemented and will always fail
execution_monitor.cpp:(.text+0x96a): warning: feenableexcept is not implemented and will always fail
make[2]: Leaving directory '/home/ubuntu/bitcoin/src'
```

laanwj commented:
@theuni thanks for review
@fanquake thanks for making the changes - will take over your branch

laanwj committed:
depends: Mention RISC-V known compilation issue with gcc-7.3.x

laanwj commented:
Updated branch:
- Took over @fanquake's branch that copies the autoconf artifacts instead of the big ugly patch
- Added a comment about the RISC-V compilation issue

theuni commented:
utACK 974f0bf8e684696be7796dbf3d48ff0a41f4ac26.
